mcmc/StatisticalUtils.h

```diff
@@ -0,0 +1,248 @@
+// ****************
+
+  double ADstat = std::fabs(CumulativeData - CumulativeMC)/ std::sqrt(CumulativeJoint*(1 - CumulativeJoint));
+
+  if( std::isinf(ADstat) || std::isnan(ADstat)) return 0;
+  return ADstat;
+}
+
+
+// ****************
+/// @brief KS: https://esjeevanand.uccollege.edu.in/wp-content/uploads/sites/114/2020/08/NON-PARAMTERIC-TEST-6.pdf
+inline int GetNumberOfRuns(std::vector<int> GroupClasifier) {
+// ****************
+
+  int NumberOfRuns = 0;
+  int PreviousGroup = -999;
+
+  //KS: If group changed increment run
+  for (unsigned int i = 0; i < GroupClasifier.size(); i++)
+  {
+    if(GroupClasifier[i] != PreviousGroup)
+      NumberOfRuns++;
+    PreviousGroup = GroupClasifier[i];
+  }
+
+  return NumberOfRuns;
+}
+
+// ****************
+inline double GetBetaParameter(const double data, const double mc, const double w2, TestStatistic TestStat) {
+// ****************
+  double Beta = 0.0;
+
+  if (TestStat == kDembinskiAbdelmottele) {
+    //the so-called effective count
+    const double k = mc*mc / w2;
+    //Calculate beta which is scaling factor between true and generated MC
+    Beta = (data + k) / (mc + k);
+  }
+  //KS: Below is technically only true for Cowan's BB, which will not be true for Poisson or IceCube, because why not...
+  else {
+    // CW: Barlow-Beeston uses fractional uncertainty on MC, so sqrt(sum[w^2])/mc
+    const double fractional = std::sqrt(w2)/mc;
+    // CW: -b/2a in quadratic equation
+    const double temp = mc*fractional*fractional-1;
+    // CW: b^2 - 4ac in quadratic equation
+    const double temp2 = temp*temp + 4*data*fractional*fractional;
+    if (temp2 < 0) {
+      std::cerr << "Negative square root in Barlow Beeston coefficient calculation!" << std::endl;
+      std::cerr << __FILE__ << ":" << __LINE__ << std::endl;
+      throw;
+    }
+    // CW: Solve for the positive beta
+    Beta = (-1*temp+std::sqrt(temp2))/2.;
+  }
+  return Beta;
+}
+
+
+// *********************
+/// @brief Based on http://probability.ca/jeff/ftpdir/adaptex.pdf
+inline double GetSubOptimality(std::vector<double> EigenValues, const int TotalTarameters) {
+// *********************
+  double sum_eigenvalues_squared_inv = 0.0;
+  double sum_eigenvalues_inv = 0.0;
+  for (unsigned int j = 0; j < EigenValues.size(); j++)
+  {
+    //KS: IF Eigen values are super small skip them
+    //if(EigenValues[j] < 0.0000001) continue;
+    sum_eigenvalues_squared_inv += std::pow(EigenValues[j], -2);
+    sum_eigenvalues_inv += 1.0 / EigenValues[j];
+  }
+  const double SubOptimality = TotalTarameters * sum_eigenvalues_squared_inv / std::pow(sum_eigenvalues_inv, 2);
+  return SubOptimality;
+}
+
+// *********************
+// Interquartile Range (IQR)
+inline double GetIQR(TH1D *Hist){
+// *********************
+
+  if(Hist->Integral() == 0) return 0.0;
+
+  double quartiles_x[3] = {0.25, 0.5, 0.75};
+  double quartiles[3];
+
+  Hist->GetQuantiles(3, quartiles, quartiles_x);
+
+  return quartiles[2] - quartiles[0];
+}
```

Comment:
Random question. I have only skimmed through here. If you want me to go through and check some of these checks in more detail then I can do.

Reply:
Most of these are stuff moved from T2K. All or almost all of StatisticalUtils.h were already valdiated. Here I moved them to have common statistiacl funcitons within MaCh3